README.md
# Fiber Modules

A list of modules useful for using with [fiber](https://github.com/gofiber/fiber)

## Install

```bash
go get github.com/gkampitakis/fiber-modules
```

## Contents

- [healthcheck](./healthcheck/README.md)
- [gracefulshutdown](./gracefulshutdown/README.md)
- [bodyvalidator](./bodyvalidator/README.md)

## Local Development

This is a general guide when you want to test some changes in a `go module` 
before publishing.

You want to test some changes in `./fiber-modules` in a 
`./server`.

Inside `./server`

```bash
# Assuming that your path structure is
# ../fiber-modules
# ../server
go mod edit -replace=github.com/gkampitakis/fiber-modules@v0.0.1-beta=../fiber-modules

go get -d github.com/gkampitakis/fiber-modules@v0.0.1-beta
```

With this way we replace the module with the local instance. Hope this helps.

## License

MIT License

CHANGELOG.md
## [1.1.10](https://github.com/gkampitakis/fiber-modules/compare/v1.1.9...v1.1.10) (2022-09-01)


### Bug Fixes

* **deps:** update all dependencies ([#16](https://github.com/gkampitakis/fiber-modules/issues/16)) ([5846c41](https://github.com/gkampitakis/fiber-modules/commit/5846c413c0387160b8063a2cfa485b17f705f6cd))

## [1.1.9](https://github.com/gkampitakis/fiber-modules/compare/v1.1.8...v1.1.9) (2022-08-01)


### Bug Fixes

* **deps:** update all dependencies ([#15](https://github.com/gkampitakis/fiber-modules/issues/15)) ([978de9c](https://github.com/gkampitakis/fiber-modules/commit/978de9c5060ce918647452bd4ad3edaff6d53697))

## [1.1.8](https://github.com/gkampitakis/fiber-modules/compare/v1.1.7...v1.1.8) (2022-07-01)


### Bug Fixes

* **deps:** update all dependencies ([#14](https://github.com/gkampitakis/fiber-modules/issues/14)) ([f773aef](https://github.com/gkampitakis/fiber-modules/commit/f773aefe2bd7d602e9054cd84e2112feec0996ee))

## [1.1.7](https://github.com/gkampitakis/fiber-modules/compare/v1.1.6...v1.1.7) (2022-06-06)


### Bug Fixes

* **deps:** update all dependencies ([#13](https://github.com/gkampitakis/fiber-modules/issues/13)) ([4213fe1](https://github.com/gkampitakis/fiber-modules/commit/4213fe192981610a9ea2e9b4d4a5c6b1d209e462))

## [1.1.6](https://github.com/gkampitakis/fiber-modules/compare/v1.1.5...v1.1.6) (2022-05-23)


### Bug Fixes

* **gracefulshutdown:** nil errorhandler ([#12](https://github.com/gkampitakis/fiber-modules/issues/12)) ([de6318c](https://github.com/gkampitakis/fiber-modules/commit/de6318c6fb77bc561700f3fc300de5b1b62e98b3))

## [1.1.5](https://github.com/gkampitakis/fiber-modules/compare/v1.1.4...v1.1.5) (2022-05-02)


### Bug Fixes

* **deps:** update module github.com/gofiber/fiber/v2 to v2.32.0 ([#11](https://github.com/gkampitakis/fiber-modules/issues/11)) ([1ad0d23](https://github.com/gkampitakis/fiber-modules/commit/1ad0d23d738ae58dea55fee89d0cd0b2614a45d8))

## [1.1.4](https://github.com/gkampitakis/fiber-modules/compare/v1.1.3...v1.1.4) (2022-02-01)


### Bug Fixes

* **deps:** update module github.com/gofiber/fiber/v2 to v2.25.0 ([#8](https://github.com/gkampitakis/fiber-modules/issues/8)) ([1f926b8](https://github.com/gkampitakis/fiber-modules/commit/1f926b847912c01f4e9411a0a2cb1f99387ee47a))

## [1.1.3](https://github.com/gkampitakis/fiber-modules/compare/v1.1.2...v1.1.3) (2022-01-01)


### Bug Fixes

* **deps:** update module github.com/gofiber/fiber/v2 to v2.24.0 ([#7](https://github.com/gkampitakis/fiber-modules/issues/7)) ([d912bce](https://github.com/gkampitakis/fiber-modules/commit/d912bce6018a016d4580bb6001e621af55318918))

## [1.1.2](https://github.com/gkampitakis/fiber-modules/compare/v1.1.1...v1.1.2) (2021-12-01)


### Bug Fixes

* **deps:** update module github.com/gofiber/fiber/v2 to v2.22.0 ([#6](https://github.com/gkampitakis/fiber-modules/issues/6)) ([5d6c73e](https://github.com/gkampitakis/fiber-modules/commit/5d6c73e6a26920507cbb4216f9286f0b90744920))

## [1.1.1](https://github.com/gkampitakis/fiber-modules/compare/v1.1.0...v1.1.1) (2021-11-15)


### Bug Fixes

* **bodyvalidator:** add option for overriding bad request response ([#5](https://github.com/gkampitakis/fiber-modules/issues/5)) ([06f64db](https://github.com/gkampitakis/fiber-modules/commit/06f64dbb8ccf7102d474d959d13d94ced2cd34c5))

# [1.1.0](https://github.com/gkampitakis/fiber-modules/compare/v1.0.2...v1.1.0) (2021-11-15)


### Features

* **bodyvalidator:** implement bodyvalidator module ([#4](https://github.com/gkampitakis/fiber-modules/issues/4)) ([31c5faa](https://github.com/gkampitakis/fiber-modules/commit/31c5faa9fc2d3fa1bf5e3ba676b5301e539d769d))

## [1.0.2](https://github.com/gkampitakis/fiber-modules/compare/v1.0.1...v1.0.2) (2021-11-07)


### Bug Fixes

* **gracefulshutdown:** implement gracefulshutdown module ([#3](https://github.com/gkampitakis/fiber-modules/issues/3)) ([4629686](https://github.com/gkampitakis/fiber-modules/commit/462968672fadb6f189a25d197034e2cd131934a3))

## [1.0.1](https://github.com/gkampitakis/fiber-modules/compare/v1.0.0...v1.0.1) (2021-11-07)


### Bug Fixes

* **healthcheck:** make redundant structs private ([6e523af](https://github.com/gkampitakis/fiber-modules/commit/6e523af89b133bf03704a98e7cc22b6e433a061a))

# 1.0.0 (2021-11-06)


### Features

* **healthcheck:** implement healthcheck module ([#1](https://github.com/gkampitakis/fiber-modules/issues/1)) ([29b42d0](https://github.com/gkampitakis/fiber-modules/commit/29b42d0db78ef4225b538526d5ee2e81a544b418))


healthcheck/config_test.go
package healthcheck

import (
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
)

func TestConfig(t *testing.T) {
	t.Run("Default Values", func(t *testing.T) {
		cfg := config{}

		assert.False(t, cfg.TimeoutEnabled, "timeout should be false by default")
		assert.Empty(t, cfg.TimeoutPeriod, "timeoutPeriod should be 0 by default")
		assert.Empty(t, cfg.ServiceName, "serviceName should be \"\" by default")
		assert.Nil(t, cfg.HealthChecks, "map should be nil by default")
		assert.False(t, cfg.ShowErrors, "showErrors should be false by default")
	})

	t.Run("EnableTimeout", func(t *testing.T) {
		t.Run("should enable timeout and set default period if not set", func(t *testing.T) {
			cfg := config{}
			EnableTimeout()(&cfg)

			assert.True(t, cfg.TimeoutEnabled)
			assert.Equal(t, time.Duration(30), cfg.TimeoutPeriod)
		})

		t.Run("should not override timeout period", func(t *testing.T) {
			cfg := config{}
			SetTimeoutPeriod(15)(&cfg)
			EnableTimeout()(&cfg)

			assert.True(t, cfg.TimeoutEnabled)
			assert.Equal(t, time.Duration(15), cfg.TimeoutPeriod)
		})
	})

	t.Run("ShowErrors", func(t *testing.T) {
		cfg := config{}
		ShowErrors()(&cfg)

		assert.True(t, cfg.ShowErrors, "should set show errors to true")
	})

	t.Run("SetTimeoutPeriod", func(t *testing.T) {
		t.Run("should set timeout period and auto enable timeout if false", func(t *testing.T) {
			cfg := config{}
			SetTimeoutPeriod(20)(&cfg)

			assert.True(t, cfg.TimeoutEnabled)
			assert.Equal(t, time.Duration(20), cfg.TimeoutPeriod)
		})
	})

	t.Run("RegisterHealthChecks", func(t *testing.T) {
		cfg := config{}
		checks := HealthchecksMap{"test": func() error { return nil }}

		RegisterHealthChecks(checks)(&cfg)

		assert.Equal(t, checks, cfg.HealthChecks)
	})
}


gracefulshutdown/config_test.go
package gracefulshutdown

import (
	"errors"
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
)

func TestConfig(t *testing.T) {
	t.Run("should return default values", func(t *testing.T) {
		cfg := Default()

		assert.Equal(t, time.Duration(15), cfg.Period)
		assert.True(t, cfg.Enabled)
		assert.Nil(t, cfg.ListenErrorHandler)
		assert.Nil(t, cfg.ShutdownFns)
		assert.Nil(t, cfg.Signals)
	})

	t.Run("should return default and set shutdownFns", func(t *testing.T) {
		fns := []func() error{
			func() error {
				return errors.New("mock-error")
			},
			func() error {
				return nil
			},
		}
		cfg := WithShutdownFns(fns)

		assert.Equal(t, time.Duration(15), cfg.Period)
		assert.True(t, cfg.Enabled)
		assert.Nil(t, cfg.ListenErrorHandler)
		assert.Equal(t, fns, cfg.ShutdownFns)
		assert.Nil(t, cfg.Signals)
	})
}


healthcheck/healthcheck_test.go
package healthcheck

import (
	"encoding/json"
	"errors"
	"io/ioutil"
	"net/http"
	"testing"
	"time"

	"github.com/gofiber/fiber/v2"
	"github.com/stretchr/testify/assert"
)

func healthRequest(t *testing.T, app *fiber.App, timeout int) ([]byte, int) {
	req, err := http.NewRequest(
		"GET",
		"/health",
		nil,
	)

	if err != nil {
		t.Fatal(err)
	}

	res, err := app.Test(req, timeout)

	if err != nil {
		t.Fatal(err)
	}

	body, err := ioutil.ReadAll(res.Body)

	if err != nil {
		t.Fatal(err)
	}

	return body, res.StatusCode
}

func setupHealthcheck(app *fiber.App, handler fiber.Handler) {
	app.Get("/health", handler)
}

func TestHealthcheckRoute(t *testing.T) {
	t.Run("should return just a status", func(t *testing.T) {
		app := fiber.New()
		setupHealthcheck(app, New(
			SetServiceName("mock-service"),
		))

		response := HealthCheckResponse{}

		body, statusCode := healthRequest(t, app, 100)
		err := json.Unmarshal(body, &response)

		if err != nil {
			t.Fatal(err)
		}

		assert.Equal(t, 200, statusCode)

		assert.Equal(t, "mock-service", response.Service)
		assert.Equal(t, 0, len(response.HealthChecks))
	})

	t.Run("should list all checks healthy", func(t *testing.T) {
		app := fiber.New()
		checks := HealthchecksMap{
			"check1": func() error { return nil },
			"check2": func() error { return nil },
			"check3": func() error { return nil },
		}

		setupHealthcheck(app, New(
			RegisterHealthChecks(checks),
		))
		response := HealthCheckResponse{}

		body, statusCode := healthRequest(t, app, 100)
		err := json.Unmarshal(body, &response)
		if err != nil {
			t.Fatal(err)
		}

		assert.Equal(t, 200, statusCode)
		assert.Equal(t, 3, len(response.HealthChecks))
		for label, value := range response.HealthChecks {
			assert.Equal(t, "healthy", value)
			_, exists := checks[label]
			assert.True(t, exists)
		}
	})

	t.Run("should handle panic inside checks", func(t *testing.T) {
		app := fiber.New()
		checks := HealthchecksMap{
			"check1": func() error { return nil },
			"check2": func() error {
				panic("boo 👻")
			},
			"check3": func() error { return nil },
		}

		setupHealthcheck(app, New(
			RegisterHealthChecks(checks),
		))
		response := HealthCheckResponse{}

		body, statusCode := healthRequest(t, app, 100)
		err := json.Unmarshal(body, &response)

		if err != nil {
			t.Fatal(err)
		}

		checksResults := response.HealthChecks

		assert.Equal(t, 500, statusCode)
		assert.Equal(t, 3, len(checksResults))
		assert.Equal(t, "healthy", checksResults["check1"])
		assert.Equal(t, "paniced with error: boo 👻", checksResults["check2"])
		assert.Equal(t, "healthy", checksResults["check3"])
	})

	t.Run("should report slow checks as timedout and 500 statusCode", func(t *testing.T) {
		app := fiber.New()
		checks := HealthchecksMap{
			"check1": func() error { return errors.New("mock-error") },
			"check2": func() error {
				time.Sleep(2 * time.Second)
				return nil
			},
			"check3": func() error { return nil },
		}

		setupHealthcheck(app, New(
			RegisterHealthChecks(checks),
			SetTimeoutPeriod(1),
		))
		response := HealthCheckResponse{}

		body, statusCode := healthRequest(t, app, 3000)
		err := json.Unmarshal(body, &response)
		if err != nil {
			t.Fatal(err)
		}

		checksResults := response.HealthChecks

		assert.Equal(t, 500, statusCode)
		assert.Equal(t, 3, len(checksResults))
		assert.Equal(t, "unhealthy", checksResults["check1"])
		assert.Equal(t, "Timeout after 1 seconds", checksResults["check2"])
		assert.Equal(t, "healthy", checksResults["check3"])
	})

	t.Run("should show errors from checks", func(t *testing.T) {
		app := fiber.New()
		checks := HealthchecksMap{
			"check1": func() error { return errors.New("mock-error") },
			"check2": func() error {
				time.Sleep(2 * time.Second)
				return nil
			},
			"check3": func() error { return nil },
		}

		setupHealthcheck(app, New(
			RegisterHealthChecks(checks),
			ShowErrors(),
		))

		response := HealthCheckResponse{}

		body, statusCode := healthRequest(t, app, 3000)
		err := json.Unmarshal(body, &response)
		if err != nil {
			t.Fatal(err)
		}

		checksResults := response.HealthChecks

		assert.Equal(t, 500, statusCode)
		assert.Equal(t, 3, len(checksResults))
		assert.Equal(t, "mock-error", checksResults["check1"])
		assert.Equal(t, "healthy", checksResults["check2"])
		assert.Equal(t, "healthy", checksResults["check3"])
	})
}


bodyvalidator/config.go
package bodyvalidator

import "github.com/qri-io/jsonschema"

type Config struct {
	// String literal representing json schema
	SchemaLiteral string
	// Path to json file containing json schema ( relative to root path )
	SchemaPath string
}

type Keywords map[string]jsonschema.KeyMaker

type BadRequestResponse func([]jsonschema.KeyError) interface{}

type globalConfig struct {
	customKeywords     Keywords
	exposeErrors       bool
	badRequestResponse BadRequestResponse
}

/*
Expose a description where the validation error occurred
*/
func ExposeErrors() func(*globalConfig) {
	return func(cfg *globalConfig) {
		cfg.exposeErrors = true
	}
}

/*
Register custom keywords to jsonschema (https://github.com/qri-io/jsonschema)
*/
func RegisterKeywords(keywords Keywords) func(*globalConfig) {
	return func(cfg *globalConfig) {
		cfg.customKeywords = keywords
	}
}

/*
Override default response with your custom response
*/
func SetResponse(response BadRequestResponse) func(*globalConfig) {
	return func(cfg *globalConfig) {
		cfg.badRequestResponse = response
	}
}


bodyvalidator/bodyvalidator.go
package bodyvalidator

import (
	"encoding/json"
	"errors"
	"fmt"
	"log"
	"net/http"
	"os"
	"path/filepath"

	"github.com/gofiber/fiber/v2"
	"github.com/qri-io/jsonschema"
)

var logFatal = log.Fatal

func New(cFns ...func(*globalConfig)) func(Config) fiber.Handler {
	globalCfg := &globalConfig{}

	for _, fn := range cFns {
		fn(globalCfg)
	}

	if globalCfg.badRequestResponse == nil {
		globalCfg.badRequestResponse = defaultResponse(globalCfg.exposeErrors)
	}

	// if keywords are given we need to load the draft again
	if len(globalCfg.customKeywords) > 0 {
		for prop, maker := range globalCfg.customKeywords {
			jsonschema.RegisterKeyword(prop, maker)
		}

		jsonschema.LoadDraft2019_09()
	}

	return func(cfg Config) fiber.Handler {
		v, err := loadValidator(cfg)
		if err != nil {
			logFatal(err)
		}

		return func(c *fiber.Ctx) error {
			validationErrors, err := v.ValidateBytes(c.Context(), c.Body())
			if err != nil {
				return c.Status(http.StatusBadRequest).
					JSON(map[string]interface{}{
						"message":    fmt.Sprintf("Bad request: %s", errors.Unwrap(err)),
						"statusCode": 400,
					})
			}

			if len(validationErrors) > 0 {
				return c.Status(http.StatusBadRequest).
					JSON(globalCfg.badRequestResponse(validationErrors))
			}

			return c.Next()
		}
	}
}

func loadValidator(cfg Config) (*jsonschema.Schema, error) {
	if cfg.SchemaLiteral == "" && cfg.SchemaPath == "" {
		warning("no schema provided")
		return nil, nil
	}
	schema := []byte(cfg.SchemaLiteral)

	if cfg.SchemaLiteral == "" {
		wd, err := os.Getwd()
		if err != nil {
			return nil, err
		}

		schema, err = os.ReadFile(filepath.Join(wd, cfg.SchemaPath))
		if err != nil {
			return nil, err
		}
	}

	s := new(jsonschema.Schema)
	err := json.Unmarshal(schema, &s)
	if err != nil {
		return nil, err
	}

	return s, nil
}

func defaultResponse(exposeErrors bool) BadRequestResponse {
	return func(errors []jsonschema.KeyError) interface{} {
		res := map[string]interface{}{
			"message":    "Bad request",
			"statusCode": 400,
		}

		if exposeErrors {
			res["description"] = errors
		}

		return res
	}
}

func warning(message string) {
	fmt.Printf("\033[33m[Warning]: %s \033[0m\n", message)
}


gracefulshutdown/gracefulshutdown.go
package gracefulshutdown

import (
	"log"
	"os"
	"os/signal"
	"syscall"
	"time"

	"github.com/gofiber/fiber/v2"
)

var defaultSignals = []os.Signal{
	syscall.SIGINT,
	syscall.SIGTERM,
}

/* Starts fiber server
Calls app.Listen if "Enabled" intercepts os.Signals
and provides some period for handling requests and calling closing functions
and closing connections
*/
func Listen(app *fiber.App, addr string, cfg Config) {
	errorHandler, signals := getDefaults(cfg)

	if !cfg.Enabled {
		startApp(app, addr, errorHandler)
		return
	}

	graceStartApp(
		app,
		addr,
		errorHandler,
		signals,
		cfg.Period,
		cfg.ShutdownFns,
	)
}

func getDefaults(c Config) (func(error), []os.Signal) {
	signals := defaultSignals

	if c.Signals != nil {
		signals = c.Signals
	}

	errorHandler := c.ListenErrorHandler
	if errorHandler == nil {
		errorHandler = func(err error) {
			log.Fatal(err)
		}
	}

	return errorHandler, signals
}

func startApp(app *fiber.App, addr string, errorHandler func(error)) {
	if err := app.Listen(addr); err != nil {
		errorHandler(err)
	}
}

func graceStartApp(
	app *fiber.App,
	addr string,
	errorHandler func(error),
	signals []os.Signal,
	period time.Duration,
	shutdownFns []func() error,
) {
	c := make(chan os.Signal, 1)
	signal.Notify(c, signals...)

	go func() {
		<-c

		log.Println("server gracefully shutting down")
		err := app.Shutdown()
		if err != nil {
			errorHandler(err)
		}

		log.Println("calling shutdown functions")

		for _, fn := range shutdownFns {
			go executeFn(fn)
		}
	}()

	startApp(app, addr, errorHandler)

	select {
	case <-c:
		log.Println("received second signal")
	case <-time.After(time.Second * period):
	}
}

func executeFn(fn func() error) {
	defer func() {
		if err := recover(); err != nil {
			log.Println(err)
		}
	}()

	if err := fn(); err != nil {
		log.Println(err)
	}
}


healthcheck/healthcheck.go
package healthcheck

import (
	"fmt"
	"net/http"
	"runtime"
	"time"

	"github.com/gofiber/fiber/v2"
)

var upTime = time.Now()

type HealthCheckResponse struct {
	Service      string            `json:"service,omitempty"`
	Uptime       string            `json:"uptime"`
	Memory       memoryMetrics     `json:"memory"`
	GoRoutines   int               `json:"go_routines"`
	HealthChecks map[string]string `json:"health_checks,omitempty"`
}

type memoryMetrics struct {
	ResidentSetSize  uint64 `json:"rss"`
	TotalAlloc       uint64 `json:"total_alloc"`
	HeapAlloc        uint64 `json:"heap_alloc"`
	HeapObjectsCount uint64 `json:"heap_objects_count"`
}

type checkResult struct {
	msg   string
	label string
}

func New(options ...func(*config)) fiber.Handler {
	cfg := config{}

	for _, o := range options {
		o(&cfg)
	}

	return registerHealthcheck(cfg)
}

// @Description Route reporting health of service
// @Summary Healthcheck route
// @Tags health
// @Accept text/plain
// @Product json/application
// @Success 200 {object} map[string]string "This can be dynamic and add more fields in checks"
// @Failure 500 {object} map[string]string "The route can return 500 in case of failed check,timeouts or panic"
// @Router /health [get]
func registerHealthcheck(cfg config) fiber.Handler {
	return func(ctx *fiber.Ctx) error {

		if cfg.HealthChecks == nil {
			return ctx.Status(http.StatusOK).JSON(prepareResponse(cfg.ServiceName, nil))
		}

		status := http.StatusOK
		response := prepareResponse(cfg.ServiceName, map[string]string{})
		c := make(chan checkResult)
		checksLength := len(cfg.HealthChecks)

		for label, control := range cfg.HealthChecks {
			go check(
				label,
				control,
				&status,
				c,
				cfg,
			)
		}

		for i := 0; i < checksLength; i++ {
			checkResponse := <-c
			response.HealthChecks[checkResponse.label] = checkResponse.msg
		}

		return ctx.Status(status).JSON(response)
	}
}

func prepareResponse(serviceName string, checks map[string]string) *HealthCheckResponse {
	var mem runtime.MemStats
	runtime.ReadMemStats(&mem)

	return &HealthCheckResponse{
		Uptime:  time.Since(upTime).String(),
		Service: serviceName,
		Memory: memoryMetrics{
			ResidentSetSize:  mem.HeapSys,
			TotalAlloc:       mem.TotalAlloc,
			HeapAlloc:        mem.HeapAlloc,
			HeapObjectsCount: mem.HeapObjects,
		},
		GoRoutines:   runtime.NumGoroutine(),
		HealthChecks: checks,
	}
}

func check(
	label string,
	control func() error,
	status *int,
	c chan<- checkResult,
	cfg config,
) {
	internalChan := make(chan checkResult, 1)

	go func() {
		defer func() {
			if e := recover(); e != nil {
				internalChan <- checkResult{msg: fmt.Errorf("paniced with error: %v", e).Error(), label: label}
				if *status == http.StatusOK {
					*status = http.StatusInternalServerError
				}
			}
		}()

		err := control()
		if err == nil {
			internalChan <- checkResult{msg: "healthy", label: label}
			return
		}

		if *status == http.StatusOK {
			*status = http.StatusInternalServerError
		}

		msg := "unhealthy"

		if cfg.ShowErrors {
			msg = err.Error()
		}

		internalChan <- checkResult{msg, label}
	}()

	if cfg.TimeoutEnabled {
		select {
		case tmp := <-internalChan:
			c <- tmp
		case <-time.After(time.Second * cfg.TimeoutPeriod):
			c <- checkResult{msg: fmt.Sprintf("Timeout after %d seconds", cfg.TimeoutPeriod), label: label}
		}

		return
	}

	c <- <-internalChan
}
